### PR TITLE
[PluggableDevice]Add kernel and stream executor c api extension

### PR DESCRIPTION
**Pull request by jzhoulon**
Extend current kernel C API and refactor some stream executor C API

### CODE REVIEW

**Comment by jzhoulon** [comment]
@annarev @sanjoy @penpornk This is the PR mentioned in [#43610](https://github.com/tensorflow/tensorflow/pull/43610), thanks

File: tensorflow/c/BUILD [hunk 1]
```diff
@@ -548,6 +548,7 @@ tf_cuda_library(
             ":tf_tensor",
             "//tensorflow/core:framework",
             "//tensorflow/core:framework_lite",
+	    "//tensorflow/c/experimental/stream_executor:stream_executor",
```
Comment (annarev) [inline tensorflow/c/BUILD:551]: nit. align indentation.

File: tensorflow/c/kernels.h [hunk 2]
```diff
@@ -65,6 +66,11 @@ typedef struct TF_KernelBuilder TF_KernelBuilder;
 typedef struct TF_OpKernelConstruction TF_OpKernelConstruction;
 typedef struct TF_OpKernelContext TF_OpKernelContext;
 
+// TF_InitKernel to do op/kernel registration.
+// Plugin should either implement TF_InitKernel to register kernels or use
+// static registration. This function should register all kernels in a plugin.
```
Comment (annarev) [inline tensorflow/c/kernels.h:71]: Let's remove mention of "static registration" per Sanjoy's comment on this PR: https://github.com/tensorflow/tensorflow/pull/43610/files

**Review by annarev — APPROVED** [review]

File: tensorflow/c/BUILD (hunk 1 above)
Comment (jzhoulon) [reply tensorflow/c/BUILD]: done, thanks for reminder.

File: tensorflow/c/kernels.h (hunk 2 above)
Comment (jzhoulon) [reply tensorflow/c/kernels.h]: done

**Review by annarev — APPROVED** [review]

File: tensorflow/c/experimental/stream_executor/stream_executor_internal.h [hunk 3]
```diff
@@ -37,6 +38,16 @@ port::Status InitStreamExecutorPlugin(void* dso_handle);
 // testing).
 port::Status InitStreamExecutorPlugin(SEInitPluginFn init_fn);
 
+namespace {
+struct TFStatusDeleter {
+  void operator()(TF_Status* s) const { TF_DeleteStatus(s); }
+};
+using OwnedTFStatus = std::unique_ptr<TF_Status, TFStatusDeleter>;
+}  // namespace
```
Comment (sanjoy) [inline tensorflow/c/experimental/stream_executor/stream_executor_internal.h:46]: No anonymous namespaces in headers please: https://google.github.io/styleguide/cppguide.html#Unnamed_Namespaces_and_Static_Variables

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/c/experimental/stream_executor/stream_executor_internal.h (hunk 3 above)
Comment (jzhoulon) [reply tensorflow/c/experimental/stream_executor/stream_executor_internal.h:46]: done.

**Review by fchollet — APPROVED** [review]
(For tf-api-owners)

File: tensorflow/c/experimental/stream_executor/stream_executor_internal.h (hunk 3 above)
Comment (jzhoulon) [reply tensorflow/c/experimental/stream_executor/stream_executor_internal.h]: @sanjoy I have changed the code as your request, could you help to take a look? If there is any thing need to change, please let me know. Thanks.

**Pull request merged**